packages/typespec-vscode/src/npm-utils.ts

```diff
@@ -0,0 +1,143 @@
+import fs from "fs";
+import path from "path";
+import semver from "semver";
+import logger from "./log/logger.js";
+import { ExecOutput, loadModule, spawnExecution, spawnExecutionEvents } from "./utils.js";
+
+export enum InstallationAction {
+  Install = "Install",
+  Upgrade = "Upgrade",
+  Skip = "Skip",
+  Cancel = "Cancel",
+}
+
+export enum npmDependencyType {
+  dependencies = "dependencies",
+  peerDependencies = "peerDependencies",
+  devDependencies = "devDependencies",
+}
+
+export interface NpmPackageInfo {
+  name: string;
+  version?: string;
+  resolved?: string;
+  overridden?: string;
+  dependencies?: Record<string, NpmPackageInfo>;
+}
+
+export class NpmUtil {
+  private cwd: string;
+
+  constructor(cwd: string) {
+    this.cwd = cwd;
+  }
+
+  public async npmInstallPackages(
+    packages: string[] = [],
+    options: any = {},
+    on?: spawnExecutionEvents,
+  ): Promise<ExecOutput> {
+    return spawnExecution("npm", ["install", ...packages], this.cwd, on);
+  }
+
+  public async ensureNpmPackageInstall(
+    packageName: string,
+    version?: string,
+  ): Promise<{ action: InstallationAction; version?: string }> {
+    const { installed: isPackageInstalled, version: installedVersion } =
+      await this.isPackageInstalled(packageName);
+    if (isPackageInstalled) {
+      if (version && installedVersion !== version) {
+        return {
+          action: InstallationAction.Upgrade,
+          version: version,
+        };
+      }
+      return { action: InstallationAction.Skip, version: installedVersion };
+    } else {
+      return { action: InstallationAction.Install, version: version };
```

Comment:
any reason to return version? seems you always return the version the called passed you...

Reply:
I want to return the expected version of the package in the environment

Reply:
but in the implementation, you always return the version passed in, isn't it?

Reply:
When it is downgrade, it will return the current existing version.